### PR TITLE
change cache_spec default value to True

### PR DESCRIPTION
**Pull request by bendichter**
fix https://github.com/NeurodataWithoutBorders/pynwb/pull/710

## Checklist

- [x] Have you checked our [Contributing](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst) document?
- [x] Have you ensured the PR description clearly describes problem and the solution?
- [x] Is your contribution compliant with our coding style ? This can be checked running `flake8` from the source directory.
- [x] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/hdmf-dev/hdmf/pulls) for the same change?
- [x] Have you included the relevant issue number using `#XXX` notation where `XXX` is the issue number ? By including "Fix #XXX" you allow GitHub to close the corresponding issue.

### CODE REVIEW

**Review by oruebel — DISMISSED** [review]

**Comment by t-b** [comment]
@bendichter Are the pynwb tests run against this version of hdmf? They were not passing last time I tried that.

**Comment by bendichter** [comment]
@t-b hmm let me check on this

**Comment by bendichter** [comment]
There are two issues we need to resolve before merging.

1) As @t-b pointed out, this currently causes an error in pynwb.
`python test.py -v`
```python
======================================================================
ERROR: test_append (test_io.TestAppend)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bendichter/dev/pynwb/tests/integration/test_io.py", line 274, in test_append
    io.write(nwb)
  File "/Users/bendichter/dev/hdmf/src/hdmf/utils.py", line 388, in func_call
    return func(self, **parsed['args'])
  File "/Users/bendichter/dev/hdmf/src/hdmf/backends/hdf5/h5tools.py", line 216, in write
    ns_builder.export('namespace', writer=writer)
  File "/Users/bendichter/dev/hdmf/src/hdmf/utils.py", line 388, in func_call
    return func(self, **parsed['args'])
  File "/Users/bendichter/dev/hdmf/src/hdmf/spec/write.py", line 198, in export
    writer.write_spec(out, path)
  File "/Users/bendichter/dev/hdmf/src/hdmf/backends/hdf5/h5_utils.py", line 147, in write_spec
    return self.__write(spec, path)
  File "/Users/bendichter/dev/hdmf/src/hdmf/backends/hdf5/h5_utils.py", line 143, in __write
    dset = self.__group.create_dataset(name, data=data, dtype=self.__str_type)
  File "/anaconda3/lib/python3.6/site-packages/h5py/_hl/group.py", line 139, in create_dataset
    self[name] = dset
  File "/anaconda3/lib/python3.6/site-packages/h5py/_hl/group.py", line 371, in __setitem__
    h5o.link(obj.id, self.id, name, lcpl=lcpl, lapl=self._lapl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5o.pyx", line 202, in h5py.h5o.link
RuntimeError: Unable to create link (name already exists)\
```
This error does not appear on the dev branch and I don't see why this chance would create this error, so some investigation is needed. 

2) Integration tests are now much slower because it needs to cache the spec each time. I think it would be best if we changed the pynwb integration test code to override the default for cache_spec and set it to `False` in the majority of cases, and then turn it on for a few tests where we test `cache_spec` specifically.

**Comment by bendichter** [comment]
OK 1 was handled by checking if a spec has already been written before writing it. 2 was handled by setting the default for integration tests to cache_spec=False: https://github.com/NeurodataWithoutBorders/pynwb/pull/945

**Comment by bendichter** [comment]
for full test coverage, we need to add a test for appending to a file with cache_spec=True in the original and second write

File: src/hdmf/backends/hdf5/h5_utils.py [hunk 1]
```diff
@@ -140,7 +140,10 @@ def stringify(spec):
 
     def __write(self, d, name):
         data = self.stringify(d)
-        dset = self.__group.create_dataset(name, data=data, dtype=self.__str_type)
+        if name in self.__group:  # handle append mode
```
Comment (ajtritt) [inline src/hdmf/backends/hdf5/h5_utils.py:143]: This behavior is handled by [require_dataset](http://docs.h5py.org/en/stable/high/group.html?highlight=require_dataset#Group.require_dataset)

**Review by ajtritt — CHANGES_REQUESTED** [review]

**Comment by rly** [comment]
See PR #84 for revised PR

**Pull request closed without merging**